File: apps/bnest-app-e2e/tests/steps/authentication.steps.ts
import { readFileSync } from "node:fs";
import path from "node:path";
import { expect, type BrowserContext, type Page } from "@playwright/test";
import { createBdd } from "playwright-bdd";
import {
  jsonFiles,
  login,
  runtimeDigest,
  submitInitialAccountsWithSafetyChecks,
  type InitialAccount,
} from "../support/authentication";
import {
  initialTestIdentities,
  isolatedTestIdentity,
  testIdentityForProject,
  userDataRelativePath,
  type TestIdentity,
} from "../support/test-identity";

const { Given, Then, When } = createBdd();

let browserBContext: BrowserContext | undefined;
let browserBPage: Page | undefined;
let setupSafetyChecks = {
  sawIrreversibleWarning: false,
  passwordRequirementsEnforced: false,
  noPasswordLengthRule: false,
};
let userDataBefore = "";
let activeIdentity: TestIdentity = testIdentityForProject("chromium");

Given("an approved user is logged in", async ({ page, $testInfo }) => {
  activeIdentity = isolatedTestIdentity($testInfo);
  await login(page, activeIdentity.admin);
});

Given("an approved admin is logged in", async ({ page, $testInfo }) => {
  activeIdentity = isolatedTestIdentity($testInfo);
  await login(page, activeIdentity.admin);
});
Given("an approved parent is logged in", async ({ page, $testInfo }) => {
  activeIdentity = isolatedTestIdentity($testInfo);
  await login(page, activeIdentity.parent);
});

Given("an approved child is logged in", async ({ page, $testInfo }) => {
  activeIdentity = isolatedTestIdentity($testInfo);
  await login(page, activeIdentity.child);
});

Given(
  "a visitor has no authenticated Bnest session",
  async ({ page, $testInfo }) => {
    activeIdentity = isolatedTestIdentity($testInfo);
    await page.context().clearCookies();
    userDataBefore = runtimeDigest(
      userDataRelativePath(activeIdentity.admin.username),
    );
  },
);

When(
  "the visitor opens the protected route {string}",
  async ({ page }, route: string) => {
    await page.goto(route);
  },
);

Then("Bnest redirects the visitor to login", async ({ page }) => {
  await expect(page).toHaveURL(/\/login\?return_to=/u);
});

Then("the login form replaces protected home actions", async ({ page }) => {
  await expect(page.locator("#login-form")).toBeVisible();
  await expect(page.locator("[data-role=chat-entry]")).toHaveCount(0);
  await expect(page.locator("[data-role=admin-settings-entry]")).toHaveCount(0);
  await expect(page.locator("[data-role=admin-schedules-entry]")).toHaveCount(
    0,
  );
});

Then("Bnest does not read or write user data", ({ page }) => {
  void page;
  expect(
    runtimeDigest(userDataRelativePath(activeIdentity.admin.username)),
  ).toBe(userDataBefore);
});

Given("Bnest has no bootstrap journal", async ({ page }) => {
  const response = await page.goto("/setup");
  expect(response?.status()).toBe(200);
  await expect(
    page.getByRole("heading", { name: "Create the first family accounts" }),
  ).toBeVisible();
});

When(
  "the maintainer submits all initial accounts including an administrator",
  async ({ page }) => {
    const accounts: InitialAccount[] = initialTestIdentities.flatMap(
      (identity) => [
        { ...identity.admin, role: "Parents", admin: true },
        { ...identity.child, role: "Children", admin: false },
        { ...identity.parent, role: "Parents", admin: false },
      ],
    );
    setupSafetyChecks = await submitInitialAccountsWithSafetyChecks(
      page,
      accounts,
    );
  },
);

Then(
  "Bnest warns that later account management and password recovery are unavailable",
  ({ page }) => {
    void page;
    expect(setupSafetyChecks.sawIrreversibleWarning).toBe(true);
  },
);

Then("Bnest creates the accounts exactly once", async ({ page }) => {
  await expect(page).toHaveURL(/\/login$/u);
  await expect(
    page.getByText(
      "Initial accounts created. Setup is now permanently closed.",
    ),
  ).toBeVisible();
});

Then(
  "setup and public registration are unavailable afterward",
  async ({ page }) => {
    expect((await page.request.get("/setup")).status()).toBe(404);
    expect((await page.request.get("/register")).status()).toBe(404);
  },
);

Then(
  "Bnest accepts the passwords without a character-count rule",
  ({ page }) => {
    void page;
    expect(setupSafetyChecks.noPasswordLengthRule).toBe(true);
  },
);

Then(
  "Bnest rejects a password missing a letter, number, or punctuation mark",
  ({ page }) => {
    void page;
    expect(setupSafetyChecks.passwordRequirementsEnforced).toBe(true);
  },
);

Given("an approved user account exists", async ({ page, $testInfo }) => {
  activeIdentity = isolatedTestIdentity($testInfo);
  await page.context().clearCookies();
  await page.goto("/login");
  await expect(page.locator("[data-phx-main]")).toHaveClass(/phx-connected/u);
});

Given(
  "an approved user account exists with an Argon2id verifier",
  async ({ page, $testInfo }) => {
    activeIdentity = isolatedTestIdentity($testInfo);
    await page.context().clearCookies();
    await page.goto("/login");
    await expect(page.locator("[data-phx-main]")).toHaveClass(/phx-connected/u);
  },
);

When("the user logs in with valid credentials", async ({ page }) => {
  await page.getByLabel("Username").fill(activeIdentity.admin.username);
  await page.getByLabel("Password").fill(activeIdentity.admin.password);
  await page.getByRole("button", { name: "Log in" }).click();
});

Then("the protected home page is available", async ({ page }) => {
  await expect(
    page.getByRole("heading", { name: "Beaver Nest" }),
  ).toBeVisible();
  await expect(page.locator(".home-status")).toContainText(
    activeIdentity.admin.username,
  );
});

When("the user logs out from that browser", async ({ page }) => {
  await page.getByRole("button", { name: "Log out" }).click();
});

Then("that browser must log in again", async ({ page }) => {
  await page.goto("/chat");
  await expect(page).toHaveURL(/\/login\?return_to=/u);
});

Then(
  "no plaintext password is stored, logged, or rendered",
  async ({ page }) => {
    const root = process.env["BNEST_E2E_RUNTIME_ROOT"];
    if (!root) throw new Error("Missing marked E2E runtime root");
    const accountBytes = jsonFiles(path.join(root, "system/accounts"))
      .map((file) => readFileSync(file, "utf8"))
      .join("");
    const pageContent = await page.content();
    for (const identity of initialTestIdentities) {
      expect(accountBytes).not.toContain(identity.admin.password);
      expect(accountBytes).not.toContain(identity.child.password);
      expect(pageContent).not.toContain(identity.admin.password);
      expect(pageContent).not.toContain(identity.child.password);
    }
  },
);

When(
  "the user reloads and reopens Bnest in the same browser",
  async ({ page }) => {
    await page.goto("about:blank");
    await page.goto("/");
  },
);

Then("the same browser remains authenticated", async ({ page }) => {
  await expect(page).toHaveURL(/\/$/u);
  await expect(page.locator(".home-status")).toContainText(
    activeIdentity.admin.username,
  );
});

Given(
  "one approved user is logged in on browser A and browser B",
  async ({ page, $testInfo }) => {
    activeIdentity = isolatedTestIdentity($testInfo);
    await login(page, activeIdentity.admin);
    const browser = page.context().browser();
    if (!browser) throw new Error("Browser fixture is unavailable");
    browserBContext = await browser.newContext({
      baseURL: new URL(page.url()).origin,
    });
    browserBPage = await browserBContext.newPage();
    await login(browserBPage, activeIdentity.admin);
  },
);

When("the user logs out from browser A", async ({ page }) => {
  await page.getByRole("button", { name: "Log out" }).click();
});

Then("browser A must log in again", async ({ page }) => {
  await page.goto("/");
  await expect(page).toHaveURL(/\/login/u);
});

Then("browser B remains authenticated", async ({ page }) => {
  void page;
  if (!browserBPage || !browserBContext)
    throw new Error("Browser B was not created");
  await browserBPage.goto("/");
  await expect(browserBPage.locator(".home-status")).toContainText(
    activeIdentity.admin.username,
  );
  await browserBContext.close();
  browserBContext = undefined;
  browserBPage = undefined;
});

File: apps/bnest-app/test/behaviour/steps/authentication_steps.exs
defmodule BnestApp.Behaviour.AuthenticationSteps do
  use ExBdd.StepDefinition

  import ExUnit.Assertions

  step "an approved user is logged in", context do
    context.behaviour_driver.establish_identity(context, :user)
  end

  step "an approved child is logged in", context do
    context.behaviour_driver.establish_identity(context, :child)
  end

  step("a visitor has no authenticated Bnest session", context,
    do: prepare(context, :unauthenticated)
  )

  step("the visitor opens the protected route {string}", %{args: [route]} = context,
    do: perform(context, :open_protected_route, [route])
  )

  step("Bnest redirects the visitor to login", context,
    do: outcome(context, :redirected_to_login)
  )

  step("Bnest does not read or write user data", context,
    do: outcome(context, :no_user_data_access)
  )

  step("Bnest has no bootstrap journal", context, do: prepare(context, :uninitialized))

  step("the maintainer submits all initial accounts including an administrator", context,
    do: perform(context, :bootstrap_accounts)
  )

  step("Bnest warns that later account management and password recovery are unavailable", context,
    do: outcome(context, :irreversible_warning)
  )

  step("Bnest creates the accounts exactly once", context,
    do: outcome(context, :accounts_created_once)
  )

  step("setup and public registration are unavailable afterward", context,
    do: outcome(context, :setup_closed)
  )

  step("Bnest accepts the passwords without a character-count rule", context,
    do: outcome(context, :passwords_without_length_rule)
  )

  step("Bnest rejects a password missing a letter, number, or punctuation mark", context,
    do: outcome(context, :password_requirements_enforced)
  )

  step("an approved user account exists", context, do: prepare(context, :approved_account))

  step("an approved user account exists with an Argon2id verifier", context,
    do: prepare(context, :approved_argon2_account)
  )

  step("the user logs in with valid credentials", context, do: perform(context, :login))

  step("the protected home page is available", context,
    do: outcome(context, :protected_home_available)
  )

  step("the user logs out from that browser", context,
    do: perform(context, :logout_current_browser)
  )

  step("that browser must log in again", context,
    do: outcome(context, :current_browser_logged_out)
  )

  step("no plaintext password is stored, logged, or rendered", context,
    do: outcome(context, :no_plaintext_password)
  )

  step("the user reloads and reopens Bnest in the same browser", context,
    do: perform(context, :reload_same_browser)
  )

  step("the same browser remains authenticated", context,
    do: outcome(context, :same_browser_authenticated)
  )

  step("one approved user is logged in on browser A and browser B", context,
    do: prepare(context, :two_browser_sessions)
  )

  step("the user logs out from browser A", context, do: perform(context, :logout_browser_a))

  step("browser A must log in again", context, do: outcome(context, :browser_a_logged_out))

  step("browser B remains authenticated", context, do: outcome(context, :browser_b_authenticated))

  step(
    "an approved user has the roles {string} and {string}",
    %{args: roles} = context,
    do: prepare(context, :multi_role_user, roles)
  )

  step("Bnest authorizes that user's own data operation", context,
    do: perform(context, :authorize_own_data)
  )

  step("the operation is allowed", context, do: outcome(context, :operation_allowed))

  step("an out-of-scope administration operation is denied", context,
    do: outcome(context, :administration_denied)
  )

  step("two approved users own separate Bnest data", context,
    do: prepare(context, :two_isolated_users)
  )

  step("the first user attempts the second user's data operation", context,
    do: perform(context, :cross_user_operation)
  )

  step("Bnest denies the operation before repository access", context,
    do: outcome(context, :denied_before_repository)
  )

  defp prepare(context, state, args \\ []),
    do: context.behaviour_driver.prepare_behaviour(context, state, args)

  defp perform(context, action, args \\ []),
    do: context.behaviour_driver.perform_behaviour(context, action, args)

  defp outcome(context, expected, args \\ []) do
    assert context.behaviour_driver.behaviour_outcome?(context, expected, args)
    context
  end
end
